Added "fancy settings" with bracket access and better subscription.

Thom1729 opened the pull request:
For #2.

The name could be changed. I'd be a bit leery of simply calling it “Settings”.

The `subscribe` method takes a “selector”, which maps the current state of a `FancySettings` to some value. (If `selector` is a string, it is interpreted as the name of a top-level setting.) When the `Settings` changes, if and only if the selected value changes, the given callback will be invoked with the new and old value as arguments. This is what I'd need for JS Custom, and I think it should handle most use cases.

Default values are not handled in this PR, though they should be easily compatible with this implementation.

The built-in `Settings` class supplies `None` for a nonexistent key instead of raising a `KeyError`. Should we follow suit or be stricter? Myself, I'd generally lean toward stricter, because the user can always provide a default value in `get`. This PR is loose, like the builtin.

It would be nice to be able to enumerate settings, but I don't think the Sublime API allows for this.

FichteFoll commented on st3/sublime_lib/fancy_settings.py:
```diff
@@ -0,0 +1,47 @@
+                callback(new_value, previous_value)
+                previous_value = new_value
+
+        self.settings.add_on_change(key, onchange)
```
I prefer using `functools.partial` over closures in situations like this, because closures create a new function for each invocation, whereas partial just binds a parameter. I would define `_on_change` as a private method then.

Thom1729 replied:
In this case, I think we need a closure (and a separate function instance per subscription) because we need to keep track of the last selected value.

FichteFoll replied:
True, I missed that it was being modified for some reason. Yeah, I guess we need it then.

FichteFoll commented on st3/sublime_lib/fancy_settings.py:
```diff
@@ -0,0 +1,47 @@
+    def __contains__(self, item):
+        return self.has(item)
+
+    def subscribe(self, key, selector, callback):
```
What is the purpose of `key` here? Imho we should try to take the `add_on_change` key away from the user so they don't have to bother with it. Besides, if you fall back the key to `selector` when it's a string, it quickly becomes non-unique.

Thom1729 replied:
`key` is there solely so that users can unsubscribe. Setting `key` when `selector` was a string was a mistake; I've fixed it.

Thom1729 replied:
That is a much better idea.

FichteFoll commented on st3/sublime_lib/fancy_settings.py:
```diff
@@ -0,0 +1,47 @@
+        return self.has(item)
+
+    def subscribe(self, key, selector, callback):
+        if isinstance(selector, str):
```
I suppose allowing sequences of strings as the selector and building the selector based on `selector = lambda this: {key: this[key] for key in selector}` would also be useful?

Thom1729 replied:
Implemented.

FichteFoll commented:
Raising `KeyError` for keys that the settings object does not have (`settings.has(key) is False`) should probably be added since it's the standard Python behavior. Returning `None` always would be akin to a `defaultdict`.

If defaults are provided to the instance in one way or another, the `.has` call can be skipped anyway.

Thom1729 commented:
I've implemented `KeyErrors` for `get` and `erase`.

Thom1729 commented:
Errors should now be raised in the right places.

I don't think that we can implement any of the `collections.abc` abstract classes because we don't have `__iter__`. I would love to implement that, but I have no idea how. I'm not sure it's possible over the API.

Thom1729 commented:
`subscribe` no longer takes a `key` argument. Instead, it generates a UUID, uses that as the key, and returns it for future use with `unsubscribe`.

FichteFoll commented:
I meant the ABC class to check whether an object is iterable using isinstance.

Thom1729 commented:
The behavior used here is [more general](https://stackoverflow.com/questions/1952464/in-python-how-do-i-determine-if-an-object-is-iterable). It may be that we don't need that level of generality, though.

Thom1729 commented:
In order to bring FancySettings closer in line with [the standard dict interface](https://docs.python.org/3/library/stdtypes.html#dict), I've added a few methods (`pop`, `setdefault`, and `update`) and changed some argument names from `name` to `key`. At this point, we're only "missing" the methods that can't be supported by the underlying implementation: `__len__`, `__iter__`, `clear`, `copy`, `fromkeys`, `items`, `keys`, `popitem`, and `values`, plus the comparison operators.

The `set`, `erase`, and `has` methods are not part of the standard interface and provide no additional functionality. Is there any reason to keep them?

On an unrelated note, is there a motivating example somewhere for default values? I sort of understand that they could be useful for view settings and such, but I can't think of a concrete use case.

FichteFoll commented:
You could also subclass `collections.abc.MutableMapping` and get the default implementations from that.

>The set, erase, and has methods are not part of the standard interface and provide no additional functionality. Is there any reason to keep them?

That's a good point. I say we remove them. This also more clearly separates the underlying API from our abstraction.

>On an unrelated note, is there a motivating example somewhere for default values? I sort of understand that they could be useful for view settings and such, but I can't think of a concrete use case.

Also a good point. On occasions, the default settings files haven't been loaded by the time the settings are accessed (despite the API being ready), so you can't always rely on them. I haven't been able to reliably reproduce this and I also haven't encountered it in a long while, but that could also be because I always include a default value in any `settings.get` call since I've had these annoying issues.
If that wasn't the case, we probably wouldn't need a default mapping as the bundled settings file should already have them.
For user settings, I suppose it is easy enough to specify the default when getting the setting, but the convenience of just accessing an item of the object without having to worry about whether there will always be a value seems quite useful regardless.

FichteFoll commented:
Hm, I guess `erase` actually behaves slightly different because it doesn't raise for a non-existant setting. Do you think we need it or can users just work around that by checking whether the setting exists or wrap it in a try-catch?

Thom1729 commented:
You can use `fancy_settings.pop(key, None)`.

FichteFoll commented on st3/sublime_lib/fancy_settings.py:
```diff
@@ -0,0 +1,87 @@
+            self.set(key, default)
+            return default
+
+    def update(*pairs, **kwargs):
```
This implementation of `update` is not correct.

It accepts either a **single** parameter that is a `dict` or a sequence of key-value pairs, or no positional parameter.
It then applies updates provided through kwargs.

Thom1729 replied:
Fixed.

FichteFoll commented:
LGTM for now. I'll take a deeper look eventually when I find time to test out all of this, but merge away.
